File: Best Time to Buy and Sell Stock/solve.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int len = prices.size();
        int buyPrice = prices[0];
        int profit = 0;
        for (int i = 1; i < len; i++)
        {

            if (prices[i - 1] < buyPrice)
                buyPrice = prices[i - 1];

            int sellPrice = prices[i];

            profit = (max(profit, sellPrice - buyPrice));
        }

        return profit;
    }
};

File: Best Time to Buy and Sell Stock/solve.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
    let minPrice = prices[0];
    const minPrices = [null];

    for (let i = 1; i < prices.length; i++) {
        minPrice = Math.min(minPrice, prices[i - 1]);
        minPrices[i] = minPrice;
    }
    // console.log(minPrices);
    let profit = 0;
    for (let i = 1; i < prices.length; i++) {
        profit = Math.max(profit, prices[i] - minPrices[i]);
    }
    return profit;
};
